src/nftfeed.sol

```diff
@@ -0,0 +1,148 @@
+
+pragma solidity >=0.5.15;
+
+import "ds-note/note.sol";
+import "tinlake-auth/auth.sol";
+import "tinlake-math/interest.sol";
+
+contract ShelfLike {
+    function shelf(uint loan) public view returns (address registry, uint tokenId);
+    function nftlookup(bytes32 nftID) public returns (uint loan);
+}
+
+contract PileLike {
+    function setRate(uint loan, uint rate) public;
+    function debt(uint loan) public returns (uint);
+    function pie(uint loan) public returns (uint);
+    function changeRate(uint loan, uint newRate) public;
+    function loanRates(uint loan) public returns (uint);
+    function file(bytes32, uint, uint) public;
+    function rates(uint rate) public view returns(uint, uint, uint ,uint48);
+}
+
+contract BaseNFTFeed is DSNote, Auth, Interest {
+    // nftID => nftValues
+    mapping (bytes32 => uint) public nftValues;
+    // nftID => risk
+    mapping (bytes32 => uint) public risk;
+    // risk => thresholdRatio
+    mapping (uint => uint) public thresholdRatio;
+    // risk => ceilingRatio
+    mapping (uint => uint) public ceilingRatio;
+
+    PileLike pile;
+    ShelfLike shelf;
+
+    constructor () public {
+        wards[msg.sender] = 1;
+    }
+
+    /// sets the dependency to another contract
+    function depend(bytes32 contractName, address addr) external auth {
+        if (contractName == "pile") {pile = PileLike(addr);}
+        else if (contractName == "shelf") { shelf = ShelfLike(addr); }
+        else revert();
+    }
+
+    /// returns a unique id based on registry and tokenId
+    /// the nftID allows to define a risk group and an nft value
+    /// before a loan is issued
+    function nftID(address registry, uint tokenId) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(registry, tokenId));
+    }
+
+    function nftID(uint loan) public view returns (bytes32) {
+        (address registry, uint tokenId) = shelf.shelf(loan);
+        return nftID(registry, tokenId);
+    }
+
+    /// Admin -- Updates
+    function setRiskGroup(uint risk_, uint thresholdRatio_, uint ceilingRatio_, uint rate_) public auth {
+        thresholdRatio[risk_] = thresholdRatio_;
+        ceilingRatio[risk_] = ceilingRatio_;
+        // the risk group is used as a rate id in the pile
+        pile.file("rate", risk_, rate_);
+    }
+
+    ///  -- Oracle Updates --
+
+    /// update the nft value
+    function update(bytes32 nftID_,  uint value) public auth {
+        nftValues[nftID_] = value;
+    }
+
+    /// update the nft value and change the risk group
+    function update(bytes32 nftID_, uint value, uint risk_) public auth {
+        require(thresholdRatio[risk_] != 0, "threshold for risk group not defined");
+
+        // change to new rate immediately in pile if a loan debt exists
+        uint loan = shelf.nftlookup(nftID_);
+        if (pile.pie(loan) != 0) {
+            pile.changeRate(loan, risk_);
+        }
+
+        risk[nftID_] = risk_;
+        nftValues[nftID_] = value;
+    }
```

Comment:
I tried to keep the `update` methods a short as possible. 
The rate for a loan is added in the `borrowEvent` method. Therefore it is not required to check if the current risk group of a loan matches with the rate of the loan in pile.

Reply:
Can you add a comment about this? The only case we need to handle in `update` therefore is if `pie>0`?

Reply:
I added a comment